go/database/mpt/proof_test.go

```diff
@@ -0,0 +1,341 @@
+						path: address1Path[2:50],
+						next: &Tag{"A_3", &Account{address: address1, pathLength: 14, info: AccountInfo{common.Nonce{1}, common.Balance{1}, common.Hash{0xAA}},
+							storage: &Tag{"K_1", &Branch{
+								children: Children{
+									key1Path[0]: &Tag{"K_2", &Extension{path: key1Path[1:40],
+										next: &Tag{"K_3", &Value{key: key1, length: 24, value: common.Value{0x12}}}}},
+									key2Path[0]: &Extension{path: key2Path[1:40],
+										next: &Value{key: key2, length: 24, value: common.Value{0x34}}},
+								}}},
+						}},
+					}},
+				}}},
+			address2Path[0]: &Tag{"B_1", &Branch{
+				children: Children{
+					address2Path[1]: &Tag{"B_2", &Extension{
+						path: address2Path[2:40],
+						next: &Tag{"B_3", &Account{address: address2, pathLength: 24, info: AccountInfo{common.Nonce{1}, common.Balance{1}, common.Hash{0xAA}},
+							storage: &Branch{
+								children: Children{
+									key1Path[0]: &Extension{path: key1Path[1:45],
+										next: &Value{key: key1, length: 19, value: common.Value{0x12}}},
+									key2Path[0]: &Extension{path: key2Path[1:45],
+										next: &Value{key: key2, length: 19, value: common.Value{0x34}}},
+								},
+							},
+						}},
+					}},
+				},
+			}},
+		}},
+	})
+
+	rootHandle := node.GetReadHandle()
+	rootHash, dirty := rootHandle.Get().GetHash()
+	if dirty {
+		t.Fatalf("expected node to be clean")
+	}
+	rootHandle.Release()
+
+	// create following shadow proofs
+	// 1. proof that contains only nodes for address1 and key1
+	// 2. proof that contains only nodes for address2
+	// 3. wide proof that includes both previous proofs
+	// 4. total proof that includes all nodes in the tree
+	address1Key1ProofShadow := testCreateShadowProofForLabels(t, ctxt, "R", "A_1", "A_2", "A_3", "K_1", "K_2", "K_3")
+	address2ProofShadow := testCreateShadowProofForLabels(t, ctxt, "R", "B_1", "B_2", "B_3")
+	wideProof := testCreateShadowProofForLabels(t, ctxt, "R", "A_1", "A_2", "A_3", "K_1", "K_2", "K_3", "B_1", "B_2", "B_3")
+	totalProof := testCreateShadowProof(t, ctxt, &root, node)
+
+	// make sure extracted proofs are different
+	if address1Key1ProofShadow.Equals(address2ProofShadow) {
+		t.Fatalf("address1 and address1Key1 proofs are equal")
+	}
+	if wideProof.Equals(address2ProofShadow) {
+		t.Fatalf("total and address1 proofs are equal")
+	}
+	if wideProof.Equals(address1Key1ProofShadow) {
+		t.Fatalf("total and address1Key1ProofShadow proofs are equal")
+	}
+	if wideProof.Equals(totalProof) {
+		t.Fatalf("total and wide proofs are equal")
+	}
+
+	t.Run("Extract", func(t *testing.T) {
+		// Test proofs can be extracted and that they match the shadow proofs
+		extractedProofAddress1Key1, exists := totalProof.Extract(rootHash, address1, key1)
+		if !exists {
+			t.Errorf("proof for %v %v %v not found", rootHash, address1, key1)
+		}
+		if got, want := extractedProofAddress1Key1, address1Key1ProofShadow; !got.Equals(want) {
+			t.Errorf("unexpected proof: got %v, want %v", got, want)
```

Comment:
nit: do proofs have a human readable print format? Otherwise, this output may be messy ... in particular since the content is a map, and their output is in random order afaik ...

Reply:
I added String(): https://github.com/Fantom-foundation/Carmen/pull/904/files#diff-14bcdb7e916407fed7057b15c0b1bb5cd85247c8542c941d7ad5065ae454f5b8R160